File: android/core/domain/src/main/kotlin/com/akaar/core/domain/model/Trade.kt
package com.akaar.core.domain.model

/**
 * A price negotiation on a listed product. Turn-based: only [awaiting] may
 * advance it, which removes every write-conflict class from the MVP.
 *
 * The buyer never sees the seller's minimum - only [band], which is derived
 * from the public listed price.
 */
data class Offer(
    val id: String,
    val productId: String,
    val buyerId: String,
    val sellerId: String,
    val status: OfferStatus,
    val awaiting: Party,
    val quantity: Int,
    val amount: Money,
    val listedPriceAtOffer: Money,
    val band: OfferBand,
    val bandReason: String?,
) {
    fun isMyTurn(myRole: Party): Boolean = !status.isSettled && awaiting == myRole
    val total: Money get() = amount * quantity
}

data class CustomRequest(
    val id: String,
    val productId: String?,
    val buyerId: String,
    val sellerId: String?,
    val status: CustomRequestStatus,
    val awaiting: Party,
    val quantity: Int?,
    val budgetPerUnit: Money?,
    val deadline: String?,
    val deliveryState: String?,
    val sizeNote: String? = null,
    val colorNote: String? = null,
    val materialNote: String? = null,
    val designNote: String? = null,
    val sampleRequired: Boolean = false,
    /** Advisory only. Never an automatic delivery promise. */
    val feasibility: FeasibilityVerdict = FeasibilityVerdict.Unavailable,
    val feasibilityReason: String? = null,
) {
    /** Matching cannot run without these, so they gate leaving draft. */
    val isSubmittable: Boolean
        get() = quantity != null && budgetPerUnit != null && deadline != null && deliveryState != null
}

data class Order(
    val id: String,
    val orderNo: String,
    val buyerId: String,
    val sellerId: String,
    val productId: String,
    val status: OrderStatus,
    val quantity: Int,
    val unitPrice: Money,
    val expectedDeliveryDate: String?,
) {
    val total: Money get() = unitPrice * quantity
}

/** A marketplace row. Buyer-facing, so it carries no private field. */
data class ProductSummary(
    val id: String,
    val title: Translatable,
    val listedPrice: Money?,
    val craftId: String?,
    val regionState: String?,
    val regionDistrict: String?,
    val moq: Int?,
    val leadTimeDays: Int?,
    val capacityPerCycle: Int?,
    val negotiable: Boolean,
    val customizationSupported: Boolean,
    val b2bEnabled: Boolean,
    val sellerId: String,
    val sellerName: String,
    val verificationTier: VerificationTier,
    val primaryMediaKey: String?,
) {
    /**
     * Why this product matched, in words. Buyers are shown reasons, never
     * scores - a number invites argument about the number.
     */
    fun matchReasons(quantity: Int?, daysAvailable: Int?): List<String> = buildList {
        if (quantity != null && moq != null && moq <= quantity) add("meets your $quantity-unit quantity")
        if (daysAvailable != null && leadTimeDays != null && leadTimeDays <= daysAvailable) {
            add("ships in $leadTimeDays days")
        }
        if (customizationSupported) add("accepts customisation")
    }
}

File: mobile/packages/core-domain/src/trade.ts
import {
  type CustomRequestStatus,
  FeasibilityVerdict,
  isSettled,
  type OfferBand,
  type OfferStatus,
  type OrderStatus,
  type Party,
  type VerificationTier,
} from "./enums.ts";
import { Money } from "./money.ts";
import type { Translatable } from "./translatable.ts";

/**
 * Port of the `Offer` data class in android/core/domain/.../Trade.kt.
 *
 * A price negotiation on a listed product. Turn-based: only `awaiting` may
 * advance it, which removes every write-conflict class from the MVP.
 *
 * The buyer never sees the seller's minimum - only `band`, which is derived
 * from the public listed price.
 */
export interface Offer {
  id: string;
  productId: string;
  buyerId: string;
  sellerId: string;
  status: OfferStatus;
  awaiting: Party;
  quantity: number;
  amount: Money;
  listedPriceAtOffer: Money;
  band: OfferBand;
  bandReason: string | null;
}

export function isMyTurn(offer: Offer, myRole: Party): boolean {
  return !isSettled(offer.status) && offer.awaiting === myRole;
}

export function offerTotal(offer: Offer): Money {
  return offer.amount.times(offer.quantity);
}

export interface CustomRequest {
  id: string;
  productId: string | null;
  buyerId: string;
  sellerId: string | null;
  status: CustomRequestStatus;
  awaiting: Party;
  quantity: number | null;
  budgetPerUnit: Money | null;
  deadline: string | null;
  deliveryState: string | null;
  sizeNote?: string;
  colorNote?: string;
  materialNote?: string;
  designNote?: string;
  sampleRequired?: boolean;
  /** Advisory only. Never an automatic delivery promise. */
  feasibility?: FeasibilityVerdict;
  feasibilityReason?: string;
}

export function customRequestDefaults(): Pick<CustomRequest, "sampleRequired" | "feasibility"> {
  return { sampleRequired: false, feasibility: FeasibilityVerdict.Unavailable };
}

/** Matching cannot run without these, so they gate leaving draft. */
export function isSubmittable(request: CustomRequest): boolean {
  return (
    request.quantity != null &&
    request.budgetPerUnit != null &&
    request.deadline != null &&
    request.deliveryState != null
  );
}

export interface Order {
  id: string;
  orderNo: string;
  buyerId: string;
  sellerId: string;
  productId: string;
  status: OrderStatus;
  quantity: number;
  unitPrice: Money;
  expectedDeliveryDate: string | null;
}

export function orderTotal(order: Order): Money {
  return order.unitPrice.times(order.quantity);
}

/** A marketplace row. Buyer-facing, so it carries no private field. */
export interface ProductSummary {
  id: string;
  title: Translatable;
  listedPrice: Money | null;
  craftId: string | null;
  regionState: string | null;
  regionDistrict: string | null;
  moq: number | null;
  leadTimeDays: number | null;
  capacityPerCycle: number | null;
  negotiable: boolean;
  customizationSupported: boolean;
  b2bEnabled: boolean;
  sellerId: string;
  sellerName: string;
  verificationTier: VerificationTier;
  primaryMediaKey: string | null;
}

/**
 * Why this product matched, in words. Buyers are shown reasons, never
 * scores - a number invites argument about the number.
 */
export function matchReasons(
  summary: ProductSummary,
  quantity: number | null,
  daysAvailable: number | null,
): string[] {
  const reasons: string[] = [];
  if (quantity != null && summary.moq != null && summary.moq <= quantity) {
    reasons.push(`meets your ${quantity}-unit quantity`);
  }
  if (daysAvailable != null && summary.leadTimeDays != null && summary.leadTimeDays <= daysAvailable) {
    reasons.push(`ships in ${summary.leadTimeDays} days`);
  }
  if (summary.customizationSupported) reasons.push("accepts customisation");
  return reasons;
}
